bootstrap/controllers/machine_deployment_controller.go

```diff
@@ -0,0 +1,437 @@
+		return fmt.Errorf("failed to patch: %w", err)
+	}
+
+	r.recorder.Eventf(
+		scope.MachineDeployment,
+		corev1.EventTypeWarning,
+		bootstrapv1.InPlaceUpgradeFailedEvent,
+		"MachineDeployment %q marked as in-place upgrade failed. Upgrade failed for machine %q",
+		scope.MachineDeployment.Name,
+		failedM.Name,
+	)
+	return nil
+}
+
+// createScope creates a new MachineDeploymentUpgradeScope.
+func (r *MachineDeploymentReconciler) createScope(md *clusterv1.MachineDeployment, ownedMachines []*clusterv1.Machine) (*MachineDeploymentUpgradeScope, error) {
+	patchHelper, err := patch.NewHelper(md, r.Client)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new patch helper: %w", err)
+	}
+
+	// NOTE(Hue): The reason we are checking the `Release` annotation as well is that we want to make sure
+	// we upgrade the new machines that joined after the initial upgrade process.
+	upgradeTo := md.Annotations[bootstrapv1.InPlaceUpgradeReleaseAnnotation]
+	if to, ok := md.Annotations[bootstrapv1.InPlaceUpgradeToAnnotation]; ok {
+		upgradeTo = to
+	}
+
+	return &MachineDeploymentUpgradeScope{
+		MachineDeployment: md,
+		UpgradeTo:         upgradeTo,
+		OwnedMachines:     ownedMachines,
+		PatchHelper:       patchHelper,
+	}, nil
+}
+
+// getCluster gets the Cluster object for the MachineDeployment.
+func (r *MachineDeploymentReconciler) getCluster(ctx context.Context, md *clusterv1.MachineDeployment) (*clusterv1.Cluster, error) {
+	cluster := &clusterv1.Cluster{}
+	clusterKey := client.ObjectKey{
+		Namespace: md.Namespace,
+		Name:      md.Spec.ClusterName,
+	}
+	if err := r.Get(ctx, clusterKey, cluster); err != nil {
+		return nil, fmt.Errorf("failed to get object: %w", err)
+	}
+
+	return cluster, nil
+}
+
+// getOwnedMachines gets the machines owned by the MachineDeployment.
+func (r *MachineDeploymentReconciler) getOwnedMachines(ctx context.Context, md *clusterv1.MachineDeployment) ([]*clusterv1.Machine, error) {
+	cluster, err := r.getCluster(ctx, md)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get cluster: %w", err)
+	}
+
+	// NOTE(Hue): The machines are not owned by the MachineDeployment directly, but by the MachineSet.
+	var (
+		msList   clusterv1.MachineSetList
+		selector = map[string]string{
+			clusterv1.ClusterNameLabel:           cluster.Name,
+			clusterv1.MachineDeploymentNameLabel: md.Name,
+		}
+	)
+	if err := r.List(ctx, &msList, client.InNamespace(cluster.Namespace), client.MatchingLabels(selector)); err != nil {
+		return nil, fmt.Errorf("failed to get MachineSetList: %w", err)
+	}
+
+	var ms clusterv1.MachineSet
+	// NOTE(Hue): The nosec is due to a false positive: https://stackoverflow.com/questions/62446118/implicit-memory-aliasing-in-for-loop
+	for _, _ms := range msList.Items { // #nosec G601
+		if util.IsOwnedByObject(&_ms, md) {
+			ms = _ms
+		}
+	}
+
+	ownedMachinesM, err := r.machineGetter.GetMachinesForCluster(ctx, client.ObjectKeyFromObject(cluster), collections.OwnedMachines(&ms))
```

Comment:
what does the M stand for?

Reply:
It stands for map since the `collections.Machines`  is basically a `map[string]*clusterv1.Machine`. I can change it if you don't like it.